### PR TITLE
fix #16898 #17621

### LINKED ISSUE #16898: lent parameter: invalid C codegen
While trying to work around #16897 I got the compiler to generate invalid C with lent input parameters.

Test case

```Nim
type
  Fp[N: static int, T] = object
    big: array[N, T]

type
  QuadraticExt* = concept x
    ## Quadratic Extension concept (like complex)
    type BaseField = auto
    x.c0 is BaseField
    x.c1 is BaseField

{.experimental:"views".}

func prod(r: var QuadraticExt, a, b: lent QuadraticExt) =
  discard

type
  Fp2[N: static int, T] = object
    c0, c1: Fp[N, T]

# This should be passed by reference,
# but concepts do not respect the 24 bytes rule
# or `byref` pragma.
var r, a, b: Fp2[6, uint64]

prod(r, a, b)

```

The proc is defined with pointers:
```Nim
/* section: NIM_merge_PROCS */
N_LIB_PRIVATE N_NIMCALL(void, prod__j3ijsYk5xsFR79akOCPoMqg)(tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* r, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* a, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* b) {
}
```
but is called with values
```Nim
N_LIB_PRIVATE N_NIMCALL(void, NimMainModule)(void) {
{
	prod__j3ijsYk5xsFR79akOCPoMqg((&r__mSc0P1JKql3o5d2mQ9bx9cMQ), a__IakjUEvgThFFu4dR6kEN0A, b__4KxaakBDXQrcdXChd3FntQ);
}
}
```
Compiler error
```
[...]/nimcache/concept_param/@mconcept_param.nim.c: In function ‘NimMainModule’:
[...]/nimcache/concept_param/@mconcept_param.nim.c:99:64: error: incompatible type for argument 2 of ‘prod__j3ijsYk5xsFR79akOCPoMqg’
   99 |  prod__j3ijsYk5xsFR79akOCPoMqg((&r__mSc0P1JKql3o5d2mQ9bx9cMQ), a__IakjUEvgThFFu4dR6kEN0A, b__4KxaakBDXQrcdXChd3FntQ);
      |                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                |
      |                                                                tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ
[...]/nimcache/concept_param/@mconcept_param.nim.c:56:143: note: expected ‘tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ *’ but argument is of type ‘tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ’
   56 | ALL(void, prod__j3ijsYk5xsFR79akOCPoMqg)(tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* r, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* a, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* b) {
      |                                                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^

[...]/nimcache/concept_param/@mconcept_param.nim.c:99:91: error: incompatible type for argument 3 of ‘prod__j3ijsYk5xsFR79akOCPoMqg’
   99 |  prod__j3ijsYk5xsFR79akOCPoMqg((&r__mSc0P1JKql3o5d2mQ9bx9cMQ), a__IakjUEvgThFFu4dR6kEN0A, b__4KxaakBDXQrcdXChd3FntQ);
      |                                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                                           |
      |                                                                                           tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ
[...]/nimcache/concept_param/@mconcept_param.nim.c:56:185: note: expected ‘tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ *’ but argument is of type ‘tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ’
   56 | oMqg)(tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* r, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* a, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* b) {
      |                                                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^

Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -O3 -fno-strict-aliasing -fno-ident   -I'[...]/.choosenim/toolchains/nim-#devel/lib' -I[...]/build -o [...]/nimcache/concept_param/@mconcept_param.nim.c.o [...]/nimcache/concept_param/@mconcept_param.nim.c' failed with exit code: 1
```

### LINKED ISSUE #17621: [View types] `lent` parameter can cause a segfault
### Example

```nim
{.experimental: "views".}

type Test = ref object
  foo: int

proc modify(self: lent Test) =
  self.foo += 1

let test = Test(foo: 12)
modify(test)
```

### Current Output

```
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Segmentation fault (core dumped)
```

### Expected Output

A compiler error. Either something like
```
Error: invalid type: 'lent Test' in this context: 'proc (self: lent Test)' for proc
```
(which is what happens without `{.experimental: "views".}`, based on https://github.com/nim-lang/Nim/issues/16898#issuecomment-770827928) or an error because `self` is modified through a `lent`.

### Possible Solution

* Disallow `lent` for parameters.

### Additional Information

Related issue: #16898.
I decided to open a new issue for this, since this causes a segfault instead of invalid codegen, but the underlying issue seems to be the same.

> $ nim -v
> Nim Compiler Version 1.5.1 [Linux: amd64]
> Compiled at 2021-04-02
> Copyright (c) 2006-2021 by Andreas Rumpf
>
> git hash: 9abd383a2ae36e50ee23b78654e9d8154db46c84
> active boot switches: -d:release

### PR DESCRIPTION
**Pull request by ringabout**
fix #16898 
fix #17621

### CODE REVIEW

File: compiler/semtypes.nim [hunk 1]
```diff
@@ -1226,9 +1227,13 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
...
       # TODO: Disallow typed/untyped in procs in the compiler/stdlib
-      if kind == skProc and (typ.kind == tyTyped or typ.kind == tyUntyped):
-        if not isMagic(getCurrOwner(c)):
-          localError(c.config, a[^2].info, "'" & typ.sym.name.s & "' is only allowed in templates and macros or magic procs")
+      if kind in {skProc, skFunc}:
+        if (typ.kind == tyTyped or typ.kind == tyUntyped) and not isMagic(getCurrOwner(c)):
+          localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
+                  "' is only allowed in templates and macros or magic procs")
+        elif typ.kind == tyLent:
+          localError(c.config, a[^2].info, "`lent` is only allowed in return type")
```
Comment (konsumlamm) [inline compiler/semtypes.nim]:
```suggestion
          localError(c.config, a[^2].info, "'lent' is only allowed in the return type")
```

Comment (konsumlamm) [reply compiler/semtypes.nim]: Ditto for the tests.

Comment (ringabout) [reply compiler/semtypes.nim]: thanks

**Comment by Araq** [comment]
Usually we use the `typeAllowed` family of functions for these "type grammar" checks.

File: compiler/semtypes.nim [hunk 2]
```diff
@@ -1226,9 +1227,10 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
     if hasType:
       typ = semParamType(c, a[^2], constraint)
       # TODO: Disallow typed/untyped in procs in the compiler/stdlib
-      if kind == skProc and (typ.kind == tyTyped or typ.kind == tyUntyped):
-        if not isMagic(getCurrOwner(c)):
-          localError(c.config, a[^2].info, "'" & typ.sym.name.s & "' is only allowed in templates and macros or magic procs")
+      if kind in {skProc, skFunc} and (typ.kind == tyTyped or typ.kind == tyUntyped) and not isMagic(getCurrOwner(c)):
+        localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
+                  "' is only allowed in templates and macros or magic procs")
```
Comment (ringabout) [inline compiler/semtypes.nim]:
```suggestion
      if kind in {skProc, skFunc} and (typ.kind == tyTyped or typ.kind == tyUntyped):
        if not isMagic(getCurrOwner(c)):
          localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
                  "' is only allowed in templates and macros or magic procs")
```

File: compiler/semtypes.nim [hunk 3]
```diff
@@ -1226,9 +1227,11 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
     if hasType:
       typ = semParamType(c, a[^2], constraint)
       # TODO: Disallow typed/untyped in procs in the compiler/stdlib
-      if kind == skProc and (typ.kind == tyTyped or typ.kind == tyUntyped):
+      if kind in {skProc, skFunc} and (typ.kind == tyTyped or typ.kind == tyUntyped):
         if not isMagic(getCurrOwner(c)):
-          localError(c.config, a[^2].info, "'" & typ.sym.name.s & "' is only allowed in templates and macros or magic procs")
+          localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
+                  "' is only allowed in templates and macros or magic procs")
```
Comment (ringabout) [inline compiler/semtypes.nim]:
```suggestion
          localError(c.config, a[^2].info, "'" & typ.sym.name.s & "' is only allowed in templates and macros or magic procs")
```

File: compiler/semtypes.nim [hunk 4]
```diff
@@ -1226,10 +1227,11 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
     if hasType:
       typ = semParamType(c, a[^2], constraint)
       # TODO: Disallow typed/untyped in procs in the compiler/stdlib
-      if kind == skProc and (typ.kind == tyTyped or typ.kind == tyUntyped):
+      if kind in {skProc, skFunc} and (typ.kind == tyTyped or typ.kind == tyUntyped):
```
Comment (ringabout) [inline compiler/semtypes.nim:1230]:
```suggestion
      if kind in {skProc, skFunc} and typ.kind in {tyTyped, tyUntyped}:
```

Comment (ringabout) [reply compiler/semtypes.nim:1230]: I can defer the change of this file because other places have the same problem: missing `skFunc`

**Pull request merged**